File: public/code/indicators__triple_confluence.py
"""
Triple Confluence Strategy
==========================
EMA + RSI + MACD üçlü onayı.

LONG koşulları (hepsi gerekli):
  - Fiyat EMA50 > EMA200 (uptrend)
  - RSI 14 > 50 (momentum yukarı) ve < 70 (aşırı alım değil)
  - MACD histogram > 0 ve MACD line > signal line

SHORT: tersleri.
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
import pandas as pd
import numpy as np
from common import make_signal, ema, rsi, macd, atr


def detect_triple_confluence_signal(
    df: pd.DataFrame,
    ema_fast: int = 50,
    ema_slow: int = 200,
    rsi_period: int = 14,
) -> dict:
    if len(df) < ema_slow + 5:
        return make_signal(reason="Insufficient data")

    ema_f = ema(df["close"], ema_fast)
    ema_s = ema(df["close"], ema_slow)
    rsi_v = rsi(df["close"], rsi_period)
    macd_l, sig_l, hist = macd(df["close"])
    atr_v = atr(df, 14)

    current = df.iloc[-1]
    current_atr = atr_v.iloc[-1]

    long_cond = (
        ema_f.iloc[-1] > ema_s.iloc[-1] and
        current["close"] > ema_f.iloc[-1] and
        50 < rsi_v.iloc[-1] < 70 and
        hist.iloc[-1] > 0 and
        macd_l.iloc[-1] > sig_l.iloc[-1] and
        hist.iloc[-1] > hist.iloc[-2]  # histogram artıyor
    )

    short_cond = (
        ema_f.iloc[-1] < ema_s.iloc[-1] and
        current["close"] < ema_f.iloc[-1] and
        30 < rsi_v.iloc[-1] < 50 and
        hist.iloc[-1] < 0 and
        macd_l.iloc[-1] < sig_l.iloc[-1] and
        hist.iloc[-1] < hist.iloc[-2]
    )

    if long_cond:
        entry = current["close"]
        stop_loss = entry - 2 * current_atr
        risk = entry - stop_loss
        return make_signal(
            signal="long",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry + risk * 1.5, entry + risk * 2.5, entry + risk * 4],
            confidence=0.8,
            reason="Triple confluence LONG (EMA+RSI+MACD)",
            metadata={"rsi": float(rsi_v.iloc[-1]), "macd_hist": float(hist.iloc[-1])},
        )

    if short_cond:
        entry = current["close"]
        stop_loss = entry + 2 * current_atr
        risk = stop_loss - entry
        return make_signal(
            signal="short",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry - risk * 1.5, entry - risk * 2.5, entry - risk * 4],
            confidence=0.8,
            reason="Triple confluence SHORT (EMA+RSI+MACD)",
            metadata={"rsi": float(rsi_v.iloc[-1]), "macd_hist": float(hist.iloc[-1])},
        )

    return make_signal(reason="No confluence")

File: public/code/indicators__triple_confluence.ts
/**
 * Triple Confluence Strategy - TypeScript
 */
import { Candle, Signal, makeSignal, ema, rsi, macd, atr } from "../../common";

export function detectTripleConfluenceSignal(
  candles: Candle[],
  opts: { emaFast?: number; emaSlow?: number; rsiPeriod?: number } = {}
): Signal {
  const { emaFast = 50, emaSlow = 200, rsiPeriod = 14 } = opts;
  if (candles.length < emaSlow + 5) return makeSignal({ reason: "Insufficient data" });

  const closes = candles.map((c) => c.close);
  const emaF = ema(closes, emaFast);
  const emaS = ema(closes, emaSlow);
  const rsiV = rsi(closes, rsiPeriod);
  const { macd: macdL, signal: sigL, histogram: hist } = macd(closes);
  const atrV = atr(candles, 14);
  const current = candles[candles.length - 1];
  const i = candles.length - 1;

  const longCond =
    emaF[i] > emaS[i] &&
    current.close > emaF[i] &&
    rsiV[i] > 50 && rsiV[i] < 70 &&
    hist[i] > 0 && macdL[i] > sigL[i] && hist[i] > hist[i - 1];

  const shortCond =
    emaF[i] < emaS[i] &&
    current.close < emaF[i] &&
    rsiV[i] > 30 && rsiV[i] < 50 &&
    hist[i] < 0 && macdL[i] < sigL[i] && hist[i] < hist[i - 1];

  if (longCond) {
    const entry = current.close;
    const stopLoss = entry - 2 * atrV[i];
    const risk = entry - stopLoss;
    return makeSignal({
      signal: "long", entry, stop_loss: stopLoss,
      take_profit: [entry + risk * 1.5, entry + risk * 2.5, entry + risk * 4],
      confidence: 0.8, reason: "Triple confluence LONG",
      timestamp: current.timestamp,
      metadata: { rsi: rsiV[i], macd_hist: hist[i] },
    });
  }
  if (shortCond) {
    const entry = current.close;
    const stopLoss = entry + 2 * atrV[i];
    const risk = stopLoss - entry;
    return makeSignal({
      signal: "short", entry, stop_loss: stopLoss,
      take_profit: [entry - risk * 1.5, entry - risk * 2.5, entry - risk * 4],
      confidence: 0.8, reason: "Triple confluence SHORT",
      timestamp: current.timestamp,
      metadata: { rsi: rsiV[i], macd_hist: hist[i] },
    });
  }
  return makeSignal({ reason: "No confluence" });
}
